File: codes/DSA 2023/infix_to_postfix.cpp
#include <iostream>
#include <stack>
using namespace std;

int precedence(char c){
    if( c=='^' || c== '$')
        return 3;
    else if( c=='*' || c== '/')
        return 2;
    else if( c=='+' || c== '-')
        return 1;
    else  
        return -1;
}
void infix_to_postfix(string s){
    stack <char> st;
    for(int i=0; i<s.length();i++){
         
        if((s[i]>='a'&& s[i]<='z') || (s[i]>='A'&& s[i]<='Z') ){
         cout<<s[i];
        } 
        else if(s[i] =='(')
         st.push('(');

        
        else if(s[i] ==')'){
            while(!st.empty()&&st.top()!= '('){
                cout<<st.top();
                st.pop();
            }
            if(!st.empty())
            st.pop();
        }
        else {
            while(!st.empty() && precedence(s[i]) <= precedence(st.top())){
                cout<<st.top();
                st.pop();
            }
            st.push(s[i]);

        }
        

    }
    while(!st.empty()){
        cout<<st.top();
        st.pop();
    }
}
int main(){
   infix_to_postfix("(a-b/c)*(a/k-l)");
 
     return 0;
}

File: codes/DSA 2023/infix_to_postfix.py
def precedence(c):
    if c == '^' or c == '$':
        return 3
    
    elif c == '*' or c == '/':
        return 2

    elif c == '+' or c == '-':
        return 1

    else :
        return -1
    
def infix_to_postfix(s):
    arr =[]
    stck =[]

    for i in range(len(s)):
        if ( s[i] >= 'a' and s[i] <= 'z') or (s[i] >= 'A' and s[i] <='Z'):
            arr.append(s[i])
        
        elif(s[i] == '(' ):
            stck.append('(')

        elif(s[i] == ')' ):
            while (len(stck) != 0) and (stck[len(stck)-1] != '('):
                arr.append(stck[len(stck)-1])
                stck.pop()
            if len(stck) != 0 :
                stck.pop() # popping the (

        else :
            while len(stck) !=0 and precedence(s[i]) <= precedence(stck[len(stck)-1]):
                arr.append(stck[len(stck)-1])
                stck.pop()
            stck.append(s[i])

    while len(stck)!=0:
        arr.append(stck[len(stck)-1])
        stck.pop()

    for i in arr:
        print(i, end ="")

infix_to_postfix('(a-b/c)*(a/k-l)')